### PR TITLE
OPERATOR-459 Migrate portworx daemonset pods

### PR DESCRIPTION
**Pull request by piyush-nimbalkar**
Signed-off-by: Piyush Nimbalkar <pnimbalkar@purestorage.com>

**What this PR does / why we need it**:

*Does handle:*
- Migration of portworx daemonset pods, assuming daemonset pods were in healthy state already
- Allows skipping nodes if user knows the issue and is okay with it 
- Handles restarts of operator during migration
- Handles daemonset placement constraints
- Logs states and errors

*Does not:*
- Does not raise events on failures
- Does not update status of the StorageCluster
- Does not migrate components
- No provision for rollback

**Which issue(s) this PR fixes** (optional)
Closes # https://portworx.atlassian.net/browse/OPERATOR-459

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/migration/migration.go [hunk 1]
```diff
@@ -486,6 +757,105 @@ func (h *Handler) findPortworxDaemonSet() (*appsv1.DaemonSet, error) {
...
+	addMigrationConstraints(&ds.Spec.Template.Spec)
+	// Set the update strategy to OnDelete to avoid restart of the daemonset
+	// pods due to the node affinity changes
+	ds.Spec.UpdateStrategy.Type = appsv1.OnDeleteDaemonSetStrategyType
+	return h.client.Update(context.TODO(), ds, &client.UpdateOptions{})
+}
+
+func addMigrationConstraints(podSpec *v1.PodSpec) {
+	if podSpec.Affinity == nil || podSpec.Affinity.NodeAffinity == nil {
+		podSpec.Affinity = &v1.Affinity{
```
Comment (pure-jliao) [inline pkg/migration/migration.go:790]: same here.

**Review by pure-jliao — APPROVED** [review]
looks pretty good to me, have some minor questions.

File: pkg/controller/storagecluster/storagecluster.go [hunk 2]
```diff
@@ -247,8 +247,9 @@ func (c *Controller) Reconcile(ctx context.Context, request reconcile.Request) (
 }
 
 func (c *Controller) waitingForMigrationApproval(cluster *corev1.StorageCluster) bool {
-	approved, err := strconv.ParseBool(cluster.Annotations[constants.AnnotationMigrationApproved])
-	return err == nil && !approved
+	_, migrationLabelPresent := cluster.Annotations[constants.AnnotationMigrationApproved]
+	return (migrationLabelPresent && cluster.Status.Phase == "") ||
```
Comment (haibinxie) [inline pkg/controller/storagecluster/storagecluster.go:251]: isn't cluster status waitingForApproval derived from the annotation, why do we need this change?

File: pkg/controller/storagecluster/storagecluster.go [hunk 3]
```diff
@@ -1028,11 +1029,45 @@ func (c *Controller) newSimulationPod(
...
+	addMigrationConstraints(&newPod.Spec)
+
 	// Add default tolerations for StorageCluster pods
 	k8s.AddOrUpdateStoragePodTolerations(&newPod.Spec)
 	return newPod, nil
 }
 
+func addMigrationConstraints(podSpec *v1.PodSpec) {
+	if podSpec.Affinity == nil || podSpec.Affinity.NodeAffinity == nil {
+		podSpec.Affinity = &v1.Affinity{
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply pkg/controller/storagecluster/storagecluster.go:1043]: we probably only has node affinity now, but would be good to handle it

File: pkg/migration/migration.go [hunk 4]
```diff
@@ -101,15 +113,195 @@ func (h *Handler) createStorageClusterIfAbsent(ds *appsv1.DaemonSet) (*corev1.St
...
+		if err := h.waitForPortworxPod(cluster, node.Name, nodeLog); err != nil {
+			return err
+		}
+
+		if err := h.markMigrationAsDone(node); err != nil {
+			return err
+		}
+
+		value = node.Labels[constants.LabelPortworxDaemonsetMigration]
+		if value == constants.LabelValueMigrationDone {
```
Comment (haibinxie) [inline pkg/migration/migration.go:181]: nit: to simplify replace if/else with one line to log with the label.

File: pkg/migration/migration.go [hunk 5]
```diff
@@ -101,15 +113,195 @@ func (h *Handler) createStorageClusterIfAbsent(ds *appsv1.DaemonSet) (*corev1.St
...
+	nodeLog *logrus.Entry,
+) error {
+	return wait.PollImmediate(podWaitInterval, daemonSetPodTerminationTimeout, func() (bool, error) {
+		node := &v1.Node{}
+		if err := h.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node); err != nil {
+			nodeLog.Errorf("Failed to get node. %v", err)
+			return false, nil
+		}
+		value := node.Labels[constants.LabelPortworxDaemonsetMigration]
+		if value == constants.LabelValueMigrationSkip {
```
Comment (haibinxie) [inline pkg/migration/migration.go:213]: when we get here, it is probably too late to skip as it would be scaled down anyway. we may just check skip in the entry and not check here, and if it happens it would migrate anyway, what do you think?

File: pkg/controller/storagecluster/storagecluster.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (piyush-nimbalkar) [reply pkg/controller/storagecluster/storagecluster.go:1043]: Did not add it because we just have node affinity right now. But I will change this for future proofing, just in case we add more to placement constraints.

File: pkg/controller/storagecluster/storagecluster.go (hunk 2 above)
Comment (piyush-nimbalkar) [reply pkg/controller/storagecluster/storagecluster.go:251]: There was a race condition where the operator reconcile loop was starting to deploy px pods before the migration handler could add the node labels to prevent the operator from starting px pods right away. Although the pods immediately get terminated but there are a lot of pods created and terminated for few seconds.

This makes sure that the operator reconcile loop waits for the migration to start and only then it is starts allowing the StorageCluster waiting to be migrated.

File: pkg/migration/migration.go (hunk 5 above)
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:213]: Actually skip does not skip the migration on that node. The daemonset pod is still deleted and the operator pod created. I don't think we need to support a case where user wants to keep some daemonset pods and some operator pods.
I added the skip to allow users to skip a node if it is not coming up for some known issue or may be an existing issue with their cluster. We don't want to get stuck at that node, nor do we want to proceed without user acknowledging the problem is known and we can skip the node. 
Basically the skip will allow the migration to proceed to a new node without worrying about whether pod comes up healthy or not.

Now the reason we want to skip from here is that if the user adds the skip flag later, we end up waiting for the timeout and then on the next restart we honor the skip. To make it react slightly faster, I added the skip check here.

**Review by haibinxie — APPROVED** [review]

File: pkg/migration/migration.go (hunk 5 above)
Comment (haibinxie) [reply pkg/migration/migration.go:213]: i was thinking about a scenario that when it is being migrated and user later added the skip annotation, we now have two pods as we dont wait here.

File: pkg/migration/migration.go [hunk 6]
```diff
@@ -101,15 +113,195 @@ func (h *Handler) createStorageClusterIfAbsent(ds *appsv1.DaemonSet) (*corev1.St
...
+			continue
+		}
+
+		if err := h.markMigrationAsStarting(node); err != nil {
+			return err
+		}
+
+		// Wait for daemonset pod to terminate, else it causes conflicts with
+		// the operator managed portworx pod
+		if err := h.waitForDaemonSetPodTermination(ds, node.Name, nodeLog); err != nil {
```
Comment (haibinxie) [inline pkg/migration/migration.go:163]: where do we let operator managed pod wait for portworx pod to be deleted?

File: pkg/migration/migration.go (hunk 5 above)
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:213]: As I said before, the skip does not stop the daemonset from deleting and operator from creating the pod. That still happens, it just skips the waiting on that node and proceeds with the migration.
